binary_port/src/binary_message.rs

```diff
@@ -0,0 +1,217 @@
+#[cfg(test)]
+use casper_types::testing::TestRng;
+#[cfg(test)]
+use rand::Rng;
+
+use bytes::Buf;
+use tokio_util::codec::{self};
+
+use crate::error::Error;
+
+type LengthEncoding = u32;
+const LENGTH_ENCODING_SIZE_BYTES: usize = std::mem::size_of::<LengthEncoding>();
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct BinaryMessage(Vec<u8>);
+
+impl BinaryMessage {
+    pub fn new(payload: Vec<u8>) -> Self {
+        BinaryMessage(payload)
+    }
+
+    pub fn payload(&self) -> &[u8] {
+        &self.0
+    }
+
+    #[cfg(test)]
+    pub(crate) fn random(rng: &mut TestRng) -> Self {
+        let len = rng.gen_range(1..=1024);
+        let payload = std::iter::repeat_with(|| rng.gen()).take(len).collect();
+        BinaryMessage(payload)
+    }
+}
+
+pub struct BinaryMessageCodec {
+    max_message_size_bytes: u32,
+}
+
+impl BinaryMessageCodec {
+    pub fn new(max_message_size_bytes: u32) -> Self {
+        Self {
+            max_message_size_bytes,
+        }
+    }
+
+    pub fn max_message_size_bytes(&self) -> u32 {
+        self.max_message_size_bytes
+    }
+}
+
+impl codec::Encoder<BinaryMessage> for BinaryMessageCodec {
+    type Error = Error;
+
+    fn encode(
+        &mut self,
+        item: BinaryMessage,
+        dst: &mut bytes::BytesMut,
+    ) -> Result<(), Self::Error> {
+        let length = item.0.len() as LengthEncoding;
+        if length > self.max_message_size_bytes {
+            return Err(Error::RequestTooLarge {
+                allowed: self.max_message_size_bytes,
+                got: length,
+            });
+        }
+        let length_bytes = length.to_le_bytes();
+        dst.extend(length_bytes.iter().chain(item.0.iter()));
+        Ok(())
+    }
+}
+
+impl codec::Decoder for BinaryMessageCodec {
+    type Item = BinaryMessage;
+
+    type Error = Error;
+
+    fn decode(&mut self, src: &mut bytes::BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if src.len() < LENGTH_ENCODING_SIZE_BYTES {
+            // Not enough bytes to read the length.
+            return Ok(None);
+        }
+        let length = LengthEncoding::from_le_bytes([src[0], src[1], src[2], src[3]]) as usize;
```

Comment:
another nitpick, but might be a good use-case for destructuring to simplify some of the validation
```rs
let (length, payload) = if let [b1, b2, b3, b4, remainder @ ..] = &src[..] {
    let length = LengthEncoding::from_le_bytes([*b1, *b2, *b3, *b4]) as usize;
    (length, remainder.get(..length).ok_or(None)?)
} else {
    // Not enough bytes to read the length.
    return Ok(None);
};
```

Reply:
Makes sense. Update in e510428a42522817f9da38d546f56e772101369c